tools/ideabuck/src/com/facebook/buck/intellij/ideabuck/run/BuckBuildBeforeRun.java

```diff
@@ -0,0 +1,170 @@
+package com.facebook.buck.intellij.ideabuck.run;
+
+import com.facebook.buck.intellij.ideabuck.build.BuckBuildCommandHandler;
+import com.facebook.buck.intellij.ideabuck.build.BuckBuildManager;
+import com.facebook.buck.intellij.ideabuck.build.BuckCommand;
+import com.facebook.buck.intellij.ideabuck.build.BuckQueryCommandHandler;
+import com.facebook.buck.intellij.ideabuck.config.BuckModule;
+import com.facebook.buck.intellij.ideabuck.file.BuckFileUtil;
+import com.facebook.buck.intellij.ideabuck.icons.BuckIcons;
+import com.google.common.base.Function;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import com.intellij.execution.application.ApplicationConfiguration;
+import com.intellij.execution.configurations.RunConfiguration;
+import com.intellij.execution.runners.ExecutionEnvironment;
+import com.intellij.openapi.actionSystem.DataContext;
+import com.intellij.openapi.externalSystem.model.ProjectSystemId;
+import com.intellij.openapi.externalSystem.service.execution.ExternalSystemBeforeRunTask;
+import com.intellij.openapi.externalSystem.service.execution.ExternalSystemBeforeRunTaskProvider;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.util.Key;
+import com.intellij.openapi.vfs.VirtualFile;
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import javax.swing.Icon;
+import java.util.List;
+
+public class BuckBuildBeforeRun extends ExternalSystemBeforeRunTaskProvider {
+
+  private static final Key<ExternalSystemBeforeRunTask> ID = Key.create("Buck.BeforeRunTask");
+  private static final ProjectSystemId SYSTEM_ID = new ProjectSystemId("buck");
+
+  private String targetResult;
+  private Project currentProject;
+
+  public BuckBuildBeforeRun(@NotNull Project project) {
+    super(SYSTEM_ID, project, ID);
+  }
+
+  @Override
+  public String getName() {
+    return "Buck Build";
+  }
+
+  @Override
+  public String getDescription(ExternalSystemBeforeRunTask task) {
+    return "Buck Build";
+  }
+
+  @NotNull
+  @Override
+  public ExternalSystemBeforeRunTask createTask(@NotNull RunConfiguration runConfiguration) {
+    return new ExternalSystemBeforeRunTask(ID, SYSTEM_ID);
+  }
+
+  @Nullable
+  @Override
+  public Icon getIcon() {
+    return BuckIcons.BUCK_TOOL_WINDOW_ICON;
+  }
+
+  @Override
+  public boolean canExecuteTask(
+      RunConfiguration configuration, ExternalSystemBeforeRunTask beforeRunTask
+  ) {
+    if (!(configuration instanceof ApplicationConfiguration)) {
+      return false;
+    }
+    VirtualFile containingFile = ((ApplicationConfiguration) configuration).getMainClass().getContainingFile().getVirtualFile();
+    VirtualFile buckFile = BuckFileUtil.getBuckFile(containingFile);
+
+    return buckFile != null;
+  }
+
+  @Override
+  public boolean executeTask(
```

Comment:
Could you split this method into smaller methods to show what's actually happening here. Right now it requires reading the whole method.

Reply:
This made this much cleaner ty :)